[P126] Add 74HC595 Shiftregisters plugin

tonhuisman opened the pull request:
On a request from [this forum thread](https://www.letscontrolit.com/forum/viewtopic.php?f=5&t=8751)

Plugin to control from 1 to ~16~ 255 74HC595 (and compatible) shift-registers, connected in a daisy-chain, as supported by these chips. ~(More daisy-chained chips can easily be supported, but the state can't be displayed on the Devices page)~
Includes an 'un-templated' (to avoid code duplication) version of the Simsso (Timo Denk) `ShiftRegister74HC595` library, as that is tried and proven technology.

Available in `TEST_E` builds.

TODO:
- [x] Restore state from RTC memory on warm boot. (max. 16 chips * 8 pins can be (re)stored)
- [x] Documentation
- [x] Testing (volunteers/help wanted 👍) ([confirmed](https://github.com/letscontrolit/ESPEasy/pull/3917#issuecomment-1019403146))


Removed from TODO list:
- Add mapping support/command(s) for 7-segment displays (_similar_ to P073)
- Testing for 7-segment display (once implemented)

I removed this feature from the TODO list, as further investigation has revealed that most of the 7-segment displays driven by 74HC595 (and compatible) chips, are actually driven in a matrix, requiring continuous updating of the data sent to the shift registers, and while an ESP is perfectly capable of doing that as a single task, it will put too much strain on the ESP for ESPEasy to function as intended and desired, resulting in possibly dropped WiFi connection, and slow response on any other task it has to perform, while the regular tasks & rules processing probably will distort the display.
NB: This plugin _can_ still be used to control a static, 1:1 connected, 74HC595 -> single 7-segment display, as in that case the 8 outputs are then each directly connected to a segment and the decimal point.

If 7-segment displays are desired, please have a look at the TM1637/MAX7219, 4, 6 or 8 digit, 7-segment displays, as supported by [P073 - Display - 7-segment display](https://espeasy.readthedocs.io/en/latest/Plugin/P073.html#p073-page).

tonhuisman committed:
[P126] Add 74HC595 Shiftregisters plugin

tonhuisman committed:
[P126] Move plugin to TEST_E set

TD-er commented on src/src/PluginStructs/P126_data_struct.cpp:
```diff
@@ -0,0 +1,168 @@
+  uint8_t *pvalue = shift->getAll();
+
+  if (nullptr != pvalue) {
+    for (uint8_t o = offset; o < offset + size; o++) {
```
```c++
    const uint8_t last = offset + size;
    for (uint8_t o = offset; o < last; o++) {
```
Also there is not a check to see if we might overflow uint8_t (also multiplying it by 8 later on)

TD-er commented on src/src/PluginStructs/P126_data_struct.h:
```diff
@@ -0,0 +1,63 @@
+  P126_data_struct() = delete;
+  ~P126_data_struct();
+
+  bool isInitialized() {
```
This is a const function

TD-er commented on src/src/PluginStructs/P126_data_struct.h:
```diff
@@ -0,0 +1,63 @@
+  uint32_t getChannelState(uint8_t offset,
+                           uint8_t size);
+
+  bool     validChannel(uint channel) {
```
This is a const function

TD-er commented on lib/ShiftRegister74HC595_NonTemplate/src/ShiftRegister74HC595_NonTemplate.cpp:
```diff
@@ -0,0 +1,102 @@
+
+// Returns the state of the given pin.
+// Either HIGH (1) or LOW (0)
+uint8_t ShiftRegister74HC595_NonTemplate::get(const uint8_t pin) {
```
const function

TD-er commented on lib/ShiftRegister74HC595_NonTemplate/src/ShiftRegister74HC595_NonTemplate.h:
```diff
@@ -0,0 +1,40 @@
+                                   const uint8_t latchPin);
+
+  void     setAll(const uint8_t *digitalValues);
+  uint8_t* getAll();
```
Looks like every use of this function could also be const.
Of make both versions, one const and one non const.

The const version:
```c++
  const uint8_t* getAll() const;
```

With this being const, the function calling this can also be made const.

tonhuisman committed:
[P126] Apply feedback, fix bugs, add SHOW_VALUES and some commands, improve 74hcSetAll for setting multiple output pins

ghtester commented:
I am already preparing my environment for tests. I would like to use this shiftregister to control relays that should survive their status (if the shiftregister & plugin works as I assume) over ESP node unexpected reboot when GPIO pins are reset to 0.

tonhuisman commented:
I'm not sure it can work that way, as the current state can't be read from the shiftregisters, it is serial-out only, AFAICS.
May need a set of 74hc165 (not sure I got the right type, the shift-in type) to read the state, but that chip is not supported by ESPEasy (yet).

ghtester commented:
I don't need to read the status, for me it's enough to write the register so 74HC595 should be fine. But I need the shift register output to keep the written status until it's updated with another data (complete 8 bits) . I suppose the outputs are not changing when the serial data are filling the register. And the must is that the outputs are still keeping the last status even though the ESP node is warm rebooted by any reason.

tonhuisman commented:
Well, then that should be covered, as _nothing_ is written to the registers at startup of the plugin, only when using the available commands: (last updated: 2022-02-27)
```
/** Commands:
 * ShiftOut,Set,<pin>,<0|1>                       : Set a single pin on or off, and update.
 * ShiftOut,SetNoUpdate,<pin>,<0|1>               : Set a single pin on or off. Use ShiftOut,Update to set outputs.
 * ShiftOut,Update                                : Update all pin states to the registers.
 * ShiftOut,SetAll,[chip:][width:]<value>...      : Set a range of chips with values, default 32 bit values (width 4).
 * ShiftOut,SetAllNoUpdate,[chip:][width:]<value> : Ditto, without immediate update. Use ShiftOut,Update to set outputs.
 * ShiftOut,SetAllLow                             : Set all register outputs to 0/low.
 * ShiftOut,SetAllHigh                            : Set all register outputs to 1/high.
 * ShiftOut,SetChipCount,<chip count>             : Set the number of chips, without restarting the plugin. Range 1..P126_MAX_CHIP_COUNT.
 * ShiftOut,SetOffset,<chip offset>               : Set the chip offset for display. Will reflect in device configuration, but not saved.
 * ShiftOut,SetHexBin,<0|1>                       : Turn off/on the Hex or Bin Values display, reflected in device configuration, not saved.
 */
```
So you can set up separate pins using `ShiftOut,SetNoUpdate` or groups using the `ShiftOut,SetAllNoUpdate` commands and then issue a `ShiftOutUpdate` command to output the state on the pins 😃 

Be careful though to not use GPIO pins that change state during startup of the ESP, as that might trigger outputting an unexpected state. See: https://espeasy.readthedocs.io/en/latest/Reference/GPIO.html

uzi18 commented:
@tonhuisman @TD-er isn't it more elegant to add setSize(uint8_t nb) method and not resize _digitalValues everytime ?

tonhuisman commented:
> not resize _digitalValues everytime

Not sure what you are hinting at here. That vector is sized at initialization, and when setting/clearing all bits, AFAICS.

ghtester commented:
Thanks for the great news. :-)
> Be careful though to not use GPIO pins that change state during startup of the ESP, as that might trigger outputting an unexpected state.

Yes that's what I was thinking of as I am already using most of GPIOs and on ESP8266 it's a problem. It will be necessary to decide what to regret (or find a way for sharing some GPIOs).

ghtester commented:
OK, so the very first quick test - 74hcSetAllHigh  command worked OK but after reboot the all outputs are set to low...
Currently I am using D3 as Data, D7 as CLK and D8 as Latch.
Bit operation works for outputs 1,2,..
But for output 0 I get:
Command unknown: 74hcSet,0,1

tonhuisman commented:
Ah, the check looks for 1..8, but I didn't subtract 1 from the pin number provided before setting that pin. Will fix.

Would you expect to set pins `0..7` or `1..8` ? (and `8..15` or `9..16` for the second chip, etc.?)

ghtester commented:
It does not matter for me, I don't know what's better - perhaps starting from 1 would be better.

The major issue for me - it's the reset outputs to low during reboot. I need to analyze if that's due to used GPIOs or what's the reason.

tonhuisman commented:
> after reboot the all outputs are set to low...
> Currently I am using D3 as Data, D7 as CLK and D8 as Latch.

Looking at the page I shared earlier, I'd try to avoid D8 and D3, as they are not `OK` in that list, they may output some signal at boot.
If the Latch wouldn't be triggered at boot it should be fine, so connecting that to an `OK` gpio like 13(D7), that will probably help.

ghtester commented:
Thanks for the comment, I'll give it a try. I am already using OK pins for another purpose so that's why I am trying those with exclamation mark first. :-) I understand I'll probably have to change that.

ghtester commented:
Just tried to swap D7 and D8 so the Latch pin is connected to 'safe' D7, Clock to D8 and Data to D3. 
The outputs are still set to 0 during node reboot.

So then I tried to reconnect the Clock pin to 'safe' D6 instead of 'unsafe' D8 (so only Data pin is connected to 'unsafe' D3) but the output clear still happens after reboot.
Are you sure there's no shiftregister initialization (zeroing) at boot?

BTW. How can the Hex State_A-3_0, State_A - State_D be used? I thought the State_A was updated in accordance to bits set through the 74hcSet (when I have just one shiftregister) but the all Values are always 0 (until set manually by Taskvalueset).
![image](https://user-images.githubusercontent.com/31805030/150405363-143593cd-6b03-4cd7-bd5f-3953c08fd542.png)

tonhuisman committed:
[P126] Bugfixes, code optimization, hex values display in uppercase

tonhuisman commented:
> The outputs are still set to 0 during node reboot.

I spoke too soon, earlier, the registers _where_ actually reset to off, at startup, sorry 😊, but that should now be fixed. The internal buffer is reset, but the state is now _not_ sent out to the registers until a command to change them is used.

The State_A..State_D variables are used to represent the `Hex` or `Bin` pin states for up to 16 chips, that's what the 'Offset for display' setting is for. Not sure if I should suppress the regular variables, they won't handle the higher values (use `TaskRun,relays` to get the current state into those variables, it will sometimes display just `0` or `nan`, 'not a number', even though there is a sensible value in there, that has to do with storing a 32 bit `unsigned int` value in a `float`).

ghtester commented:
Great, thank you for updates! I am going to recompile my custom build again, upgrade the node firmware and test again.
It would be nice if the buffer could be reset only at cold boot (don't know if it's ever possible to achieve). The best would be to fill the internal buffer value from plugin Values (State_A...State_D, which are stored in RTC memory so preserved during warm boot).

Stupid me, but I still don't understand how to operate with the State_A...State_D values. Currently there's no relation with pin states, the values are all zero despite bits are set and **TaskRun, Relays** command is performed.
![image](https://user-images.githubusercontent.com/31805030/150421677-ad957c6a-f5fd-4e81-9b1d-94a3539594ef.png)

tonhuisman committed:
[P126] Bugfix for display range

tonhuisman commented:
> the values are all zero despite bits are set and **TaskRun, Relays** command is performed.

The current state should be reflected in the `Hex State_A ...` just above the `State_A` value. Maybe I have the bytes in the wrong order, I'll check.

> The best would be to fill the internal buffer value from plugin Values (State_A...State_D, which are stored in RAM so preserved during warm boot).

I'll investigate how to restore the values after a warm reboot.

ghtester commented:
FYI I have changed the **Number of chips** to 4, then **74hcSetAllHigh**  and  **TaskRun,Relays** and this is the result:
![image](https://user-images.githubusercontent.com/31805030/150423685-c7482c8d-dc71-4dbe-a882-3cfc5c71d7dd.png)
It looks when Number of chips = 4 it works... :-) 
**74hcSet,1,1  74hcSet,2,1 TaskRun,Relays**:
![image](https://user-images.githubusercontent.com/31805030/150424547-b77fac03-c135-4543-a408-afeac4fd15b6.png)

tonhuisman commented:
After issuing the commands `74hcset,1,1` and `74hcset,8,1` and `taskrun,relays`, (and I set the display to Bin), I have this result:

![image](https://user-images.githubusercontent.com/1922393/150423506-295cdc7b-89c1-4e3e-ba4a-74b37a0962c7.png)

The State_A variable is unexpectedly _not_ displaying the current value, so I'll have a look into that.

You are showing the `nan` value I wrote about earlier, that is also something to investigate, it could know the value is set as an unsigned 32 bit integer (I set it using the designated support function), so it could display like that. I tried to set decimals to 2 and even 5, but then it still shows 0.00(000) 😞

Edit: Number of chips is still set to 1.
Edit2: You can set the Interval > 0 on the task to automatically update the Hex/Bin State display.

ghtester commented:
> You can set the Interval > 0 on the task to automatically update the Hex/Bin State display.

FYI when I set the Interval >0, the periodical crashes encountered:
<html>
<body>
<!--StartFragment-->

Boot: | Exception (18)
-- | --
Reset Reason: | Exception
Last Action before Reboot: | PLUGIN_READ: timer, id: 9
SW WD count: | 0

<!--EndFragment-->
</body>
</html>

uzi18 commented:
@tonhuisman it was just a comment to "updated" library, the comment said: 
`Refactor ShiftRegister74HC595 to ShiftRegister74HC595_NonTemplate to enable runtime sizing` but did not find method to change size in runtime.

uzi18 commented on src/src/PluginStructs/P126_data_struct.cpp:
```diff
@@ -0,0 +1,237 @@
+  String command = parseString(string, 1);
+
+  if (command.startsWith(F("74hc"))) {
+    if (command.equals(F("74hcset")) || command.equals(F("74hcsetnoupdate"))) {
```
@tonhuisman how about such idea : 
`bool hc_update = !command.contains(F("noupdate"));`
than:
`if (command.equals(F("74hcset")) || command.equals(F("74hcsetnoupdate")))` 
we can change to  
`if (command.startsWith(F("74hcset"))`

and `shift->set(pin - 1, value, command.equals(F("74hcset")));` goes to `shift->set(pin - 1, value, hc_update);`

tonhuisman replied:
> `if (command.equals(F("74hcset")) || command.equals(F("74hcsetnoupdate")))`
> we can change to
> `if (command.startsWithF("74hcset"))`

That's not really feasible as there are also the `74hcsetall` commands. Having a single bool for `noupdate` might save a few bytes, though. I'll improve that.

uzi18 replied:
@tonhuisman idea is to save more cpu time (string compare), you can use it also for 74hcsetall

tonhuisman commented:
> @tonhuisman it was just a comment to "updated" library, the comment said: `Refactor ShiftRegister74HC595 to ShiftRegister74HC595_NonTemplate to enable runtime sizing` but did not find method to change size in runtime.

I may need to re-word that part slightly, the regular implementation is sized at compile-time, by means of a C++ template. The 'runtime sizing' I was trying to describe is in fact an 'instantiation time sizing', the object can be destroyed and re-created with a different size.
I don't expect there is a need for actual runtime sizing, you won't be able to, or need to I'd expect, hook up extra chips while the device is running. That should be a 1-time setup/configuration thing, as usual for ESPEasy, IMHO.

uzi18 commented:
> That should be a 1-time setup/configuration thing, as usual for ESPEasy, IMHO.

Right, but maybe someone would like to use your modified copy of lib :)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p126-74hc595-shiftregisters

ghtester commented:
> > You can set the Interval > 0 on the task to automatically update the Hex/Bin State display.
> 
> FYI when I set the Interval >0, the periodical crashes encountered:
> 
> Boot:	Exception (18)
> Reset Reason:	Exception
> Last Action before Reboot:	PLUGIN_READ: timer, id: 9
> SW WD count:	0

The periodical Exceptions happens at least when I set the interval to 1. Maybe it's too short value for this plugin. Then the web GUI becomes frozen quickly after the (re)boot and after a while the Exception is generated.
Increasing the Interval value to 30 helped but I did not test it for a long time. The Hex State display was updated only after the web browser page refresh (another plugins like **Environment - AHT10/20/21 [TESTING]** and **Analog input - PCF8591** are updating the Values even without the web browser window refresh).
The major functions of the plugin now seems to be working fine - for me it's important the all shiftregister bits (including the first one) can be set / cleared and the shiftregister keeps the status after the warm reboot.
Thanks for the great work! :-)

tonhuisman committed:
[Lib] ShiftRegister74HC595_NT: Add setSize() method, cleanup constructor

tonhuisman committed:
[P126] Optionally restore State on warm boot, code improvements

tonhuisman commented:
> The periodical Exceptions happens at least when I set the interval to 1. Maybe it's too short value for this plugin. Then the web GUI becomes frozen quickly after the (re)boot and after a while the Exception is generated.

I've been running for a while on Interval = 1, using an ESP8266, but have not seen crashes or UI slowdown yet.

Also added restoring the values on warm boot (configurable, on by default). **Limitation:** Only the state for up to 16 chips can be restored, and the restore starts at Offset for display, other chips are reset to all outputs 0. (Only these bytes are restored from RTC memory.)
New build should be available in a short while, after Github actions run is completed.

tonhuisman commented:
> Last Action before Reboot: PLUGIN_READ: timer, id: 9

Is the 74HC595 plugin on device task 9? (or 10, it may be off by 1... as in 0-based id 😄) or what other plugin is there?

ghtester commented:
![image](https://user-images.githubusercontent.com/31805030/150640688-fe29752f-e13b-482e-a2cb-8e4010df038a.png)

The Environment - AHT10/20/21 [TESTING] plugin has the Interval set to 5 s do maybe this can play a role...
Also I still have set the Number o chips = 4.

tonhuisman commented:
So it looks like the AHT10 plugin is causing the reboot...

Edit: If Controller 001 is Domoticz, then the Idx value should be set other than 0.

ghtester commented:
Yeah but AHT10 plugin itself works fine so the issue could be related with a bit unusual current configuration.
Controller 1 is Generic HTTP (and the target IP is unavailable now as the testing node is at another - non default - location now; I can try to disable it)
Controller 2 is Home Assistant (openHAB) MQTT and it's now disabled.

tonhuisman commented:
> set the Number o chips = 4.

Ran for an hour with that setting without issues, now switched to 255 chips, to allocate the max. amount of memory, and restarted the run just now. ~Will report here in a while to see if I can have it crash.~ See 2 messages below.

ghtester commented:
OK, thanks for your time. I can also play with the configuration a bit to discover what's the root of this issue but as the crash happen quite soon after the boot and I don't have a serial connection to this node, it's a bit complicated to change the Interval settings back from 1 to 0 or disable the plugin to get a control back...

BTW. the LCD plugin also has an Interval set to 1s so I understand it may be a bit hard and unusual settings.

Update -  I have changed the 74HC595 interval to 10s, Number of chips to 1 and enabled the IR Receive plugin (which decreased a free memory significantly). So far so good...

tonhuisman committed:
[P126] Add command 74hcSetChipCount for changing the number of active chips at runtime

tonhuisman commented:
After adding the 74hcSetChipCount command, I'm now testing with Interval 1 _and_ changing the number of chips between 16 and 255 every minute. And now we wait... (psssshhh, aahhh 🍻)

tonhuisman committed:
[P126] Correct Values display and event generation, disable Formula and Decimals, update Device parameters

ghtester commented:
I have recompiled & upgrade ESP Easy firmware from latest sources, now the 74HC595 plugin is almost perfect! Only the Values item (currently Hex State_A 1_4:) on Devices page does not autorefresh like other plugins do. It's necessary to push F5 to get it updated (and show the current Value if it was changed in the meantime) in web browser. But it's a very minor difference. 
The best improvement is the configurable buffer restore from RTC memory after reboot, it's really great! Now we have a cheap and very reliable solution for controlling relays (even a lot of them!) and I am sure it's very important for many ESP Easy related projects.
The test node was working fine for several hours after the latest upgrade, now I have decreased the Interval to 5s, still everything looks OK and so far no crash. Thanks again for the amazing work! :-)

TD-er commented:
Sounds good!
And indeed controlling lots of relays is a very good use case, perhaps should be mentioned in the documentation.

tonhuisman committed:
Merge branch 'mega' into feature/p126-74hc595-shiftregisters

tonhuisman commented:
> Only the Values item (currently Hex State_A 1_4:) on Devices page does not autorefresh like other plugins do. It's necessary to push F5 to get it updated (and show the current Value if it was changed in the meantime) in web browser.

Don't know if you have other plugins active, with a (short?) interval, or have set the interval for this plugin, but that _should_ refresh the (all) values. Though I have seen that not all browser (especially Chrome) do that nicely.

TD-er commented:
Only the first 4 values of a task are updated by the JavaScript.

ghtester commented:
Well, the Devices page looks now like this:
![image](https://user-images.githubusercontent.com/31805030/151180811-e4bc2ee8-7125-4f30-88c9-1e4d7dd27251.png)
The Interval settings were / changed to:
5s /5s     74HC595 
10s / 30s System Info
5s / 30s  AHT10/20/21 
1s / 10s  LCD2004
1s / 10s  PCF8591

But the Hex State_A 1_4: of 74HC595 still does not refresh, all other tasks are refreshing properly. Tested in Firefox now, in  Chrome, Edge with earlier Interval settings - the same result.

TD-er commented:
Ah I think this may have to do with the different formatting of that task.
How does it show up in the JSON data that's fetched?
I think since it is not a standard output format, it does not have a matching HTML id on the page.

tonhuisman commented:
I am providing the data via the `PLUGIN_WEBFORM_SHOW_VALUES` plugin event, without appending the regular (and rather useless) Values. Probably when showing these regular values again, the formatted values are also updated, or maybe we can fix something in the javascript to show this when available.

ghtester commented:
![image](https://user-images.githubusercontent.com/31805030/151191034-724d2ee1-fa98-43ff-b24d-6bb4821c8bfd.png)

tonhuisman committed:
Merge branch 'mega' into feature/p126-74hc595-shiftregisters

tonhuisman committed:
[P126] Move log before next, unprocessed, argument

tonhuisman commented:
I have _cancelled_ the original TODO for adding support for 7-segment displays. Explanation is in the first comment in this PR.

uzi18 commented:
@tonhuisman do you have any link to such multiplexed display?

tonhuisman committed:
[Docs] Update copyright year in documentation  footer

tonhuisman committed:
[P126] Minor updates and corrections while writing documentation

tonhuisman committed:
[P126] Add documentation

tonhuisman commented:
> @tonhuisman do you have any link to such multiplexed display?

Well, I just googled for this type of display, [something like this qualifies](https://create.arduino.cc/projecthub/goarray/4-digit-7-segment-shift-register-counter-b7338d)
And on [Aliexpress](https://www.aliexpress.com/item/32834757140.html)

But I still suggest to use a MAX7219 controlled display like that, much easier to use, as it does the multiplexing autonomously.

tonhuisman committed:
[P126] Remove build info from screenshot

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p126-74hc595-shiftregisters

tonhuisman committed:
[P126] Documentation improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p126-74hc595-shiftregisters

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p126-74hc595-shiftregisters

tonhuisman committed:
[P126] Log optimization

ghtester commented:
Is it necessary to make a change in my Custom.h after recent changes in source codes? I can't compile with Vagrant anymore:

   default: Indexing .pio/build/custom_IR_ESP8266_4M1M/libFrameworkArduino.a
    default: Linking .pio/build/custom_IR_ESP8266_4M1M/ESP_Easy_mega_20220212_custom_IR_ESP8266_4M1M.elf
    default: /home/vagrant/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/custom_IR_ESP8266_4M1M/src/src/Helpers_tmp/__tmpfile.cpp.o:(.text._Z10PluginInitv+0xa4): undefined reference to `Plugin_126(unsigned char, EventStruct*, String&)'
    default: collect2: error: ld returned 1 exit status
    default: *** [.pio/build/custom_IR_ESP8266_4M1M/ESP_Easy_mega_20220212_custom_IR_ESP8266_4M1M.elf] Error 1

tonhuisman commented:
The inclusion of the separate `Plugin_nnn` functions has moved to Plugin_init.h/.cpp, so no longer auto-managed by the ino to cpp conversion. So if you have a plugin enabled of which the source is not available you will see this compiler error. Once you disable the `#define USES_P126` in your Custom.h you should be fine again.

ghtester commented:
Thanks for the feedback.  The question is if I am still able to compile with PR#3917 when I disable P126 in Custom.h ?

tonhuisman commented:
> The question is if I am still able to compile with PR#3917 when I disable P126 in Custom.h ?

Well, I expect you got this error when compiling another branch of ESPEasy, that doesn't have this plugin available.
When compiling _this_ PR you ~can~ *should* enable the #define, to have it included, of course 👍

ghtester commented:
Yeah I am always compiling the main branch in first step as there's a script that (re)builds the Vagrant environment completely from scratch. In second step (to compile with a PR not yet merged) I am conencting to already running Vagrant virtual machine and starting a compilation there (with PR parameter). 
So I did the same now, ignored the 1st step compilation error and the second step finished succesfully. Thanks for your advice.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p126-74hc595-shiftregisters

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p126-74hc595-shiftregisters

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p126-74hc595-shiftregisters

tonhuisman committed:
[P126] Rename command-prefix from 74hc to 74hc595, and update documentation

tonhuisman commented:
@ghtester I have changed the commands for this plugin to now be 'prefixed' with `74hc595` instead of just `74hc` as I've added a PR for 74HC165, where I used `74hc165` as command prefix, to better distinguish these commands.

tonhuisman committed:
[P126] Add output selection option, improve values display

tonhuisman committed:
[P126] Update documentation

uzi18 commented:
@tonhuisman maybe shiftin and shiftout instead of 74hc...

tonhuisman commented:
> @tonhuisman maybe shiftin and shiftout instead of 74hc...

Very good suggestion, I will change both plugins, P126 to `shiftout,<subcommand>,<args>...` and P129 to `shiftin,<subcommand>,<args>...`

ghtester commented:
Well, just another idea (when there's another change pending yet) - there are existing commands GPIO and PCFGPIO, maybe it would be nice to keep GPIO in the new commands (related to shift I/O pins) as well. What do you think?

tonhuisman commented:
> there are existing commands GPIO and PCFGPIO, maybe it would be nice to keep GPIO in the new commands (related to shift I/O pins) as well.

Well, there are a few reasons I'm less fond of that idea. GPIO stands for General Purpose Input Output, and that's not what these are, they are either output or input, so IMHO that wouldn't actually match the 'Input Output' part. And having commands like `shiftgpioout`, `shiftgpioin` or `shiftgpo` and `shiftgpi` are quite hard to read, where `shiftin` and `shiftout` are much better comprehendible, short and to the point, having short command names is also a plus in my book.

uzi18 commented:
@tonhuisman Asked for change cmd name because there are available also others shift registers than 595 and 165 you can use here. Maybe should mention that earlier but it still waiting to merge.

tonhuisman commented:
> there are available also others shift registers than 595 and 165 you can use

Yes, that is the main reason I agree with your proposed command rename 😃, and the second reason the commands being better readable.
I mention a few other chips in the documentation for this plugin. Have to investigate and include that for P129 as well.

ghtester commented:
OK, I understand, you are right. I also prefer a short command names. Hopefully the command name changes will be final. ;-)

tonhuisman committed:
[P126] Rename command to ShiftOut, and promote rest of command to second argument

tonhuisman committed:
[P126] Update documentation using new command name

tonhuisman commented:
> Hopefully the command name changes will be final. ;-)

It is nicely in line with many other plugins command handling now, shouldn't need to change again.

The only change I'm now thinking of is the name of the plugin, currently it's `Output - 74HC595 Shiftregisters`, but maybe here also the chip name should be removed.
The only issue I have with that is the name is then quite unspecific, `Output - Shift registers` (adding a space as well, to use correct English). Then P129 should also be renamed to `Input - Shift registers`, making these plugins more difficult to distinguish.
Probably, less is more, in this case 🤔

tonhuisman committed:
[P126] Rename plugin title, update documentation other minor doc. improvements

tonhuisman committed:
[P126] Always show formatted output, even when plugin is not active.

tonhuisman committed:
Merge branch 'mega' into feature/p126-74hc595-shiftregisters

TD-er committed:
Merge branch 'mega' into feature/p126-74hc595-shiftregisters
